### PR TITLE
[compiler] Implement support for hoisted and recursive functions

### PR DESCRIPTION
**Pull request by mvitousek**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #30922

Summary:
Introduces a new binding kind for functions that allows them to be hoisted. Also has the result of causing all nested function declarations to be outputted as function declarations, not as let bindings.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts [hunk 1]
```diff
@@ -1001,6 +1001,20 @@ function codegenTerminal(
...
+            loc: iterableItem.loc,
+            suggestions: null,
+          });
+        case InstructionKind.Function:
+          CompilerError.invariant(false, {
+            reason: 'Unexpected Function variable in for..in collection',
+            description: null,
+            loc: iterableItem.loc,
+            suggestions: null,
+          });
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts:1017]: we could just collapse all these unsupported cases

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts [hunk 2]
```diff
@@ -1103,6 +1117,20 @@ function codegenTerminal(
...
+        case InstructionKind.Function:
+          CompilerError.invariant(false, {
+            reason: 'Unexpected Function variable in for..of collection',
+            description: null,
+            loc: iterableItem.loc,
+            suggestions: null,
+          });
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts:1133]: same here

**Review by josephsavona — APPROVED** [review]
wow awesome!

**Comment by Mishap111** [comment]
Byte

**Pull request merged**